zerver/lib/message.py

```diff
@@ -374,6 +374,13 @@ def finalize_payload(
         if not skip_copy:
             obj = copy.copy(obj)
 
+        if obj["sender_email_address_visibility"] != UserProfile.EMAIL_ADDRESS_VISIBILITY_EVERYONE:
+            # If email address of the sender is only available to administrators,
+            # clients cannot compute gravatars, so we force-set it to false.
+            # If we plumbed the current user's role, we could allow client_gravatar=True
+            # here if the current user's role has access to the target user's email address.
+            client_gravatar = False
```

Comment:
This is messy, but I think probably the best solution available. In the future, I think we're going to want to just move off Gravatar to our own procedural generator that is a function of immutable details like the user ID and realm ID, or perhaps a random seed, so I'm not sure it's worth cleaning this up really.